### PR TITLE
BUG: nonexistent Timestamp pre-summer/winter DST w/dateutil timezone

### LINKED ISSUE #31043: BUG: nonexistent Timestamp pre-summer/winter DST change with dateutil timezone
#### Code Sample, a copy-pastable example if possible

This is fine:
```
>>> pd.__version__
'0.26.0.dev0+1790.gdd94e0db9'
>>> epoch =  1552211999999999871
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999871-0800', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999871
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999871-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552211999999999871
```
This is also fine:
```
>>> epoch =  1552212000000000000
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 03:00:00-0700', tz='dateutil/US/Pacific')
>>>
>>> t.value
1552212000000000000
>>> pd.Timestamp(t)
Timestamp('2019-03-10 03:00:00-0700', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552212000000000000
```
Meanwhile, this breaks representation and gets us nonexistent times:
```
>>> epoch =  1552211999999999872
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999872-0700', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999872
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999872-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552208399999999872
```
And right on the cusp, the value breaks too:
```
>>> epoch =  1552211999999999999
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999999-0700', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999999
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999999-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552208399999999999
```

#### Problem description

When we use `dateutil` timezones and try to create a Timestamp that is right on the cusp of the change from winter to summer time, we can get nonexistent times (the clock is supposed to jump from 2 A.M. to 3 A.M. and yet we get 2:59:59).

I've investigated this, and it appears that at 128 nanoseconds before the clock jump, DST offset and utcoffset in `dateutil` change, so we end up in a situation when the offsets are what they are supposed to be after the jump, but the time hasn't jumped yet, so the constructor returns a nonexistent time. Calling the constructor again moves the clock 1 hour back.

This can be checked out with:
```
>>> epoch =  1552211999999999872
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t.tz.dst(t)
datetime.timedelta(seconds=3600)
```

My assumption is that when we need to determine UTC offset, rounding happens at some point, and we round to `epoch=1552212000000000000`, get offset, and then use it on time pre-clock jump.

I'd like to try to fix this one.

#### Expected Output
```
>>> epoch =  1552211999999999872
>>> t = pd.Timestamp(epoch, tz='dateutil/US/Pacific')
>>> t
Timestamp('2019-03-10 01:59:59.999999872-0800', tz='dateutil/US/Pacific')
>>> t.value
1552211999999999872
>>> pd.Timestamp(t)
Timestamp('2019-03-10 01:59:59.999999872-0800', tz='dateutil/US/Pacific')
>>> pd.Timestamp(t).value
1552208399999999872
```

#### Notes

This was thought to be part of #24329 but turned to be a separate bug as I worked on closing that issue in PR #30995.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : dd94e0db9556f35a3ea91ce85714c0b1e151a770
python           : 3.7.6.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : ru_RU.UTF-8
LOCALE           : None.None

pandas           : 0.26.0.dev0+1790.gdd94e0db9
numpy            : 1.17.4
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 44.0.0.post20200106
Cython           : 0.29.14
pytest           : 5.3.2
hypothesis       : 5.1.5
sphinx           : 2.3.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.7
lxml.etree       : 4.4.2
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.1
fastparquet      : 0.3.2
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : 3.1.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.1
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pytest           : 5.3.2
s3fs             : 0.4.0
scipy            : 1.3.1
sqlalchemy       : 1.3.12
tables           : 3.6.1
tabulate         : 0.8.6
xarray           : 0.14.1
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.7
numba            : 0.47.0

</details>

### PR DESCRIPTION
**Pull request by AlexKirko**
- [X] closes #31043
- [X] tests added 1 / passed 1
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

This implements rounding down to microseconds into Timedelta.total_seconds(). Lack of this rounding led to `dateutil.tz.tzinfo.utcoffset` and `dst` breaking less than 128 nanoseconds before winter/summer DST switch. This happened due to an unintended cast to float in `Timedelta.total_seconds()` compounded with `Timedelta.value` being np.int64 type which doesn't support long arithmetic. The loss of precision led to rounding up, which meant that `total_seconds` since epoch time implied DST time while the `Timedelta.value` hasn't yet reached DST.

Details and code examples below.

<details>

This was quite a journey, but I found out what's going on. Let's say we have a `dateutil.tz.tzinfo` object named `du_tz` and want to find out the DST-aware UTC offset.

1. We call `du_tz.utcoffset(dt)` which calls `du_tz._find_ttinfo(dt)` which calls `du_tz._resolve_ambiguous_time(dt)` to find the index of the last transition time that it uses to return the correct offset.
2. `du_tz._resolve_ambiguous_time(dt)` calls `du_tz._find_last_transition(dt)` which calls the `_datetime_to_timestamp(dt)` dateutil function.
3. This is what this function does:

```
def _datetime_to_timestamp(dt):
    """
    Convert a :class:`datetime.datetime` object to an epoch timestamp in
    seconds since January 1, 1970, ignoring the time zone.
    """
    return (dt.replace(tzinfo=None) - EPOCH).total_seconds()
```

The problem is dateutil's reliance on `Timedelta.total_seconds` which casts to float:

```
    def total_seconds(self):
        """
        Total duration of timedelta in seconds (to ns precision).
        """
        return self.value / 1e9
```

Demonstration:

```
IN:
import datetime
import pandas as pd

epoch =  1552211999999999872
ts = pd.Timestamp(epoch, tz='dateutil/US/Pacific')

EPOCH = datetime.datetime.utcfromtimestamp(0)

delta = ts.replace(tzinfo=None) - EPOCH
print(delta.value)
OUT:
1552183199999999872
IN:
print(delta.total_seconds())
OUT:
1552183200.0
IN:
print(ts.tz.dst(ts))
OUT:
datetime.timedelta(seconds=3600)
```
The same thing happens with a `pytz` timezone, only pytz relies on something else to check for DST transitions:
```
IN:
import datetime
import pandas as pd

epoch =  1552211999999999872
ts = pd.Timestamp(epoch, tz='US/Pacific')

EPOCH = datetime.datetime.utcfromtimestamp(0)

delta = ts.replace(tzinfo=None) - EPOCH
print(delta.value)
OUT:
1552183199999999872
IN:
print(delta.total_seconds())
OUT:
1552183200.0
IN:
print(ts.tz.dst(ts))
OUT:
datetime.timedelta(0)
```
So `timedelta` value is okay, but `timedelta.total_seconds` hits the precision limit of floats, and this leads to rounding and an incorrect DST offset.

</details>

### CODE REVIEW

**Review by jreback — APPROVED** [review]
lgtm. ping on green.

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 1]
```diff
@@ -861,9 +861,11 @@ cdef class _Timedelta(timedelta):
 
     def total_seconds(self):
         """
-        Total duration of timedelta in seconds (to ns precision).
+        Total duration of timedelta in seconds (to microsecond precision).
         """
-        return self.value / 1e9
+        # GH 31043
+        # Round down to microseconds to avoid confusing tzinfo.utcoffset
+        return (self.value - self.value % 1000) / 1e9
```
Comment (AlexKirko) [inline pandas/_libs/tslibs/timedeltas.pyx]: We can also fix the bug with this:
```
return int(self.value) / 1000000000
```
By casting from `numpy.int64` to `int` and spelling out the denominator as an `int`, we switch to a different arithmetic implementation which handles long arithmetic better. This introduces an additional cast though and relies on the inner implementation of python. In addition, we end up returning a non-numpy type which clashes with the rest of our code.

Comment (jreback) [reply pandas/_libs/tslibs/timedeltas.pyx]: totally fine with what you have; i think we actually do this in a couple of places already

Comment (AlexKirko) [reply pandas/_libs/tslibs/timedeltas.pyx]: Hmm. Something fishy is going on with the tests.

Comment (jreback) [reply pandas/_libs/tslibs/timedeltas.pyx]: k ping when ready

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 2]
```diff
@@ -1092,3 +1092,15 @@ def test_constructor_ambigous_dst():
     expected = ts.value
     result = Timestamp(ts).value
     assert result == expected
+
+
+def test_constructor_before_dst_switch():
```
Comment (mroeschke) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: Can we test the example in the issue here? (i.e. that `Timestamp(Timestamp(epoch_time, tz=..))` doesn't change value)

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 3]
```diff
@@ -861,9 +861,12 @@ cdef class _Timedelta(timedelta):
...
     def total_seconds(self):
         """
-        Total duration of timedelta in seconds (to ns precision).
+        Total duration of timedelta in seconds (to microsecond precision).
         """
-        return self.value / 1e9
+        # GH 31043
+        # Microseconds precision to avoid confusing tzinfo.utcoffset
+        return ((self.days * 86400 + self.seconds) * 10**6 +
+                self.microseconds) / 10**6
```
Comment (AlexKirko) [inline pandas/_libs/tslibs/timedeltas.pyx]: @jreback 
I'll need a day or two. I've tried everything, including the `datetime.timedelta` implementation (current commit), but the damn thing just keeps failing my test in all the pipelines but two.
This is especially frustrating, because the code passes the test both in my Windows and Linux development environment.
This will be hard to debug without an ability to replicate the test fail directly.

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 2 above)
Comment (AlexKirko) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: @mroeschke Sure!
But the value breaks only 1 nanosecond before the switch. Currently, I'm struggling to make everything else work on every azure pipeline (see comment below).

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 4]
```diff
@@ -1092,3 +1092,17 @@ def test_constructor_ambigous_dst():
...
+    # GH 31043
+    # Make sure that calling Timestamp constructor
+    # on time just before DST switch doesn't lead to
+    # nonexistent time
+    epoch = 1552211999999999872
+    ts = Timestamp(epoch, tz="dateutil/US/Pacific")
+    delta = ts.replace(tzinfo=None) - datetime.utcfromtimestamp(0)
+    assert delta.total_seconds() < 1552183200
```
Comment (AlexKirko) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: Expanded the test a bit to see whether we end up rounding after all on most environments or if something else is going on.

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 5]
```diff
@@ -1092,3 +1092,17 @@ def test_constructor_ambigous_dst():
...
+    # Make sure that calling Timestamp constructor
+    # on time just before DST switch doesn't lead to
+    # nonexistent time
+    epoch = 1552211999999999872
+    ts = Timestamp(epoch, tz="dateutil/US/Pacific")
+    delta = ts.replace(tzinfo=None) - datetime.utcfromtimestamp(0)
+    assert 1552183199.99 < delta.total_seconds() < 1552183200
+    expected = timedelta(seconds=0)
+    result = ts.tz.dst(ts)
+    assert result == expected
```
Comment (AlexKirko) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: Okay, we now fail only on Linux minimum versions and macOS. Let's fiddle with the implementation a couple more times.

**Comment by AlexKirko** [comment]
@jreback 
Okay, I found why this breaks on minimum versions.
Before version 2.7.0, `dateutil` didn't fully rely on `timedelta.total_seconds()`. Instead it implemented this function:
```
def _total_seconds(td):
    # Python 2.6 doesn't have a total_seconds() method on timedelta objects
    return ((td.seconds + td.days * 86400) * 1000000 +
            td.microseconds) // 1000000


_total_seconds = getattr(timedelta, 'total_seconds', _total_seconds)
```
They had this for Python 2.6 datetime support and dropped it in [this PR](https://github.com/dateutil/dateutil/pull/362).
This implementation should grab our timedelta with `getattr`, but somehow it breaks.
I set up a python 3.6 environment. The tests fail with dateutil 2.6.1 and pass with dateutil 2.7.0.

Basically, I want to know how I should proceed. I can search for a workaround around their outdated workaround by setting up a second environment for testing with dateutil 2.6.1.
Or we could change our minimum supported version of dateutil to 2.7.0 if that's possible. We don't support Python 2 in pandas, so there is really no need to support a version of dateutil this low.

**Comment by jreback** [comment]
so just change the test have different cases for >=2,7 and <2,7
we already moved to
2.6.1 as the min (for 1.0.0);

File: pandas/tests/indexes/datetimes/test_timezones.py [hunk 6]
```diff
@@ -2,10 +2,12 @@
 Tests for DatetimeIndex timezone-related methods
 """
 from datetime import date, datetime, time, timedelta, tzinfo
+from distutils.version import LooseVersion
```
Comment (AlexKirko) [inline pandas/tests/indexes/datetimes/test_timezones.py]: Need to parse version properly.

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 7]
```diff
@@ -1092,3 +1094,23 @@ def test_constructor_ambigous_dst():
     expected = ts.value
     result = Timestamp(ts).value
     assert result == expected
+
+
+@pytest.mark.xfail(
+    LooseVersion(get_distribution("python-dateutil").version) < LooseVersion("2.7.0"),
```
Comment (AlexKirko) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: @jreback 
I apologize for `get_distribution("python-dateutil").version`. It's pretty horrible.
Unfortunately `from dateutil import __version__` breaks mypy checks. And `dateutil.__version__` breaks it.

I don't know why, because mypy is completely fine with `from pandas import __version__`. Both packages have the attribute and they are `str`. I checked (this check fail has been pretty hard on my nerves). 

Tried other stuff too. Only dateutil fails:
```
$ mypy pandas/tests/scalar/timestamp/test_timestamp.py
pandas\tests\scalar\timestamp\test_timestamp.py:1: error: Module 'dateutil' has no attribute '__version__'
Found 1 error in 1 file (checked 1 source file)
(pandas-dev)
```

My current solution passes all the checks, but I would appreciate suggestions if anyone sees how the ugliness can be avoided without breaking mypy.

File: pandas/tests/scalar/timestamp/test_timestamp.py (hunk 2 above)
Comment (AlexKirko) [reply pandas/tests/scalar/timestamp/test_timestamp.py]: @mroeschke 
Added to the test. Pinging you on change.

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 8]
```diff
@@ -1092,3 +1094,23 @@ def test_constructor_ambigous_dst():
...
+    # Make sure that calling Timestamp constructor
+    # on time just before DST switch doesn't lead to
+    # nonexistent time or value change
+    # Works only with dateutil >= 2.7.0 as dateutil overrid
+    # pandas.Timedelta.total_seconds with
+    # datetime.timedelta.total_seconds before
+    ts = Timestamp(epoch, tz="dateutil/US/Pacific")
+    result = ts.tz.dst(ts)
+    expected = timedelta(seconds=0)
+    assert Timestamp(ts).value == epoch
```
Comment (AlexKirko) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: @mroeschke 
This test that the value no longer shifts. In the issue, this failed for `epoch = 1552211999999999999`

**Comment by AlexKirko** [comment]
Because `dateutil.__version__` breaks the mypy check, I'm using `pandas.compat._optional._get_version`. Seems like it was written precisely for something like this, and it's certainly a better idea than importing `pkg_resources.get_distribution`.
Once the numpy dev PR is merged, I'll merge master into my branch, and we should be green.

**Comment by AlexKirko** [comment]
@jreback 
All green, please review.

**Comment by jreback** [comment]
thanks @AlexKirko very nice

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 9]
```diff
@@ -1092,3 +1093,23 @@ def test_constructor_ambigous_dst():
...
+)
+@pytest.mark.parametrize("epoch", [1552211999999999872, 1552211999999999999])
+def test_constructor_before_dst_switch(epoch):
+    # GH 31043
+    # Make sure that calling Timestamp constructor
+    # on time just before DST switch doesn't lead to
+    # nonexistent time or value change
+    # Works only with dateutil >= 2.7.0 as dateutil overrid
+    # pandas.Timedelta.total_seconds with
+    # datetime.timedelta.total_seconds before
```
Comment (pganssle) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: This does not make sense to me - `datetime.timedelta.total_seconds` should succeed, because it's equivalent to:

```python
def total_seconds(td):
  useconds = td.days * 86400
  useconds += td.seconds 
  useconds *= 1000000
  useconds += td.microseconds
  return useconds / 1e6
```

I think there's actually a deeper issue here, which is that `td.microseconds` and `td.seconds` are *rounded* rather than *truncated*. Consider this:

```python
def to_ns(td):
  ns = td.days * 86400
  ns += td.seconds
  ns *= 1000000
  ns += td.microseconds
  ns *= 1000
  ns += td.nanoseconds
  return ns

td = timedelta(1552211999999999872, unit="ns")
print(td.value)  # 1552211999999999872
print(to_ns(td))  # 1552212000000000872
```

That seems to be the actual root cause of this issue and should probably be fixed.

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 10]
```diff
@@ -1092,3 +1093,23 @@ def test_constructor_ambigous_dst():
...
+    ts = Timestamp(epoch, tz="dateutil/US/Pacific")
```
Comment (pganssle) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: Forgot to mention earlier: you should probably use `dateutil/America/Los_Angeles`, as that is the canonical name for this zone. The `US/...` zones are symlinks for backwards compatibility.
